generators/php/text.ts

```diff
@@ -10,41 +10,45 @@
 
 // Former goog.module ID: Blockly.PHP.texts
 
+import type {Block} from '../../core/block.js';
+import type {JoinMutatorBlock} from '../../blocks/text.js';
 import {Order} from './php_generator.js';
+import type {PhpGenerator} from './php_generator.js';
 
 
-export function text(block, generator) {
+export function text(block: Block, generator: PhpGenerator): [string, Order] {
   // Text value.
   const code = generator.quote_(block.getFieldValue('TEXT'));
   return [code, Order.ATOMIC];
 };
 
-export function text_multiline(block, generator) {
+export function text_multiline(block: Block, generator: PhpGenerator): [string, Order] {
   // Text value.
   const code = generator.multiline_quote_(block.getFieldValue('TEXT'));
   const order =
       code.indexOf('.') !== -1 ? Order.STRING_CONCAT : Order.ATOMIC;
   return [code, order];
 };
 
-export function text_join(block, generator) {
+export function text_join(block: Block, generator: PhpGenerator): [string, Order] {
```

Comment:
would it work to just type `block` as `JoinMutatorBlock` to begin with? haven't thought through the implications of this, just throwing it out there.

Reply:
That would be a great idea but alas it is not possible to put a `(block: JoinMutatorBlock, …) => …` into the `phpGenerator.forBlock` dictionary because that type is not assignable to `(block: Block, …) => …` (which is what the `forBlock` dictionary is declared as containing).

I've been thinking about this issue for a while, and I think the correct approach is to for the block definitions to provide [type predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) (for at least all those block type that declare additional properties) and then have the generator functions use those type predicates to narrow the type of `block` via runtime test (e.g. `if (!isJoinMutatorBlock(block)) throw new TypeError()`, after which `block` is understood to be a `JoinMutatorBlock`).

That obviously entails providing such predicates, and possibly changing the way the blocks types are declared, so I think this is out-of-scope for this PR.